### PR TITLE
Core ML: let op_linear_quantizer_config compress gather tables

### PR DESCRIPTION
**Pull request by msluszniak**
### Summary

`op_linear_quantizer_config` is applied with `op_type_configs={"gather": None}`, so gather tables are never compressed. For a transformer that is usually right: the body is most of the weight, and the embedding table is the one tensor output quality rests on most directly.

For a sentence embedder it is backwards. The table IS the model.

| model | fp16 | int8, tables exempt | int8, tables included |
|---|---|---|---|
| all-MiniLM-L6-v2 | 45.5 MB | 34.7 MB (-24%) | 22.8 MB (-50%) |
| paraphrase-multilingual-MiniLM-L12-v2 | 235.6 MB | 214.1 MB (-9%) | 118.4 MB (-50%) |

The multilingual model keeps 96 MB of fp16 vocab table it does not need. Its 250k rows are most of the file, so the current exemption leaves almost the whole win on the table.

This adds `quantize_gather_tables` to `generate_compile_specs`, off by default, so the caller who can measure the result decides. Models lowered before this spec existed are unaffected, including the tied-embedding handling from #21969.

### Scope: every float gather table, not only embeddings

The opt-in names every gather whose table is a float constant. Core ML lowers `index_select` and plain constant indexing to `gather` as well, so a float buffer such as a rotary position cache is compressed too. Confirmed on a module with no embedding layer: both caches were picked up.

Narrowing to real embedding lookups is not possible here. By the time the MIL program exists a gather is a gather, with nothing marking which came from `nn.Embedding`, so narrowing would be a shape heuristic that is wrong in a different way. The name, both helper docstrings and the partitioner docs page all say what it actually covers.

Integer tables are skipped by the opt-in, which checks the dtype. Run against the quantizer directly, a two-dimensional integer table raises and takes the conversion down, while a one-dimensional one is skipped by the quantizer itself.

Only single-axis `gather` is covered, by the opt-in and by the default exemption alike. `gather_along_axis` (`torch.gather`) and `gather_nd` (indexing with two index tensors) are compressed either way; widening the exemption would shrink what existing models compress, so it belongs in its own change.

### Does it cost anything at runtime

Measured on an iPhone 16, all-MiniLM-L6-v2, CPU_ONLY, release build, 3 warmup + 20 timed, re-taken after the float-table selection landed; the only change since is the rename.

| variant | pipeline | peak RAM | file |
|---|---|---|---|
| fp16 | 1.26 ms | 125.02 MB | 45.5 MB |
| int8, tables exempt | 1.273 ms | 113.05 MB | 34.7 MB |
| int8, tables included | 1.274 ms | 102.55 MB | 22.8 MB |

Latency is flat across all three, so compressing the tables costs no speed on the CPU compute unit, and peak memory falls 18%.

Cosine against the torch fp32 reference over 8 sentences: including the tables costs 0.00001 on all-MiniLM (0.99889 -> 0.99888) and 0.00002 on paraphrase-L12 (0.99979 -> 0.99977).

### Test plan

Four tests in `test_coreml_partitioner.py`: the compile spec round-trips and defaults to off, a caller passing `pass_names` positionally still binds it there, an embedding-only model lowers smaller with the opt-in, and a model that also indexes a uint8 buffer still lowers and still shrinks, which pins the float-table filter. That buffer is two-dimensional on purpose: a 1-D table is skipped by the quantizer on its own, so it would pass with the filter deleted. Confirmed it reaches MIL as `gather` over a `uint8` (4096, 8) constant, uncast.

cc @kimishpatel @YifanShenSZ @cymbalrush @metascroy @shoumikhin

### CODE REVIEW

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 1]
```diff
@@ -990,4 +1037,4 @@ def preprocess_multimethod(  # noqa: C901
 
         finally:
             # Clean up temporary directory
-            shutil.rmtree(str(temp_dir))
+            shutil.rmtree(str(temp_dir))
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py]: The file lost its final newline, and the format check is now red on this change. Please add the newline back at the end of the file. Running the repo formatter on the file fixes it.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 2]
```diff
@@ -396,6 +423,7 @@ def generate_compile_specs(
         compute_precision: ct.precision = ct.precision.FLOAT16,
         model_type: MODEL_TYPE = MODEL_TYPE.MODEL,
         op_linear_quantizer_config: Optional[Dict] = None,
+        quantize_embedding_tables: bool = False,
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py]: `quantize_embedding_tables` takes the old positional slot for `pass_names`. Existing six-argument calls now treat the pass list as this flag and silently use the default pass pipeline. Please move the new option after `pass_names` and add a positional compatibility test.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 3]
```diff
@@ -615,12 +649,25 @@ def _convert_to_mlmodel(
...
+            #
+            # quantize_embedding_tables drops the opt-out entirely, for models whose table
+            # is most of their weight and which have measured that compressing it is worth
+            # it. It stays off by default: the table is the one weight an embedding model's
+            # output quality rests on most directly, so opting in belongs with the caller
+            # who can measure the result.
+            if CoreMLBackend.quantize_embedding_tables_from_compile_specs(
+                compile_specs
+            ):
+                config = cto.coreml.OptimizationConfig(
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py]: `quantize_embedding_tables` removes the exemption for every `gather`, including non-floating lookup tables. With a large `uint8` table, Core ML Tools tries to quantize it and aborts conversion. Please keep unsupported gather constants exempt.

Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py]: `quantize_embedding_tables=True` removes the exclusion for every Core ML `gather`, not only embeddings. Core ML lowers `index_select` to `gather`, so this can also quantize unrelated lookup tables and change their values. Please limit the override to embedding operations, or rename and document the option as applying to every gather table.

**Review by shoumikhin — COMMENTED** [review]
Some notes inline.

**Review by shoumikhin — COMMENTED** [review]
Some notes inline.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 4]
```diff
@@ -397,9 +456,13 @@ def generate_compile_specs(
         model_type: MODEL_TYPE = MODEL_TYPE.MODEL,
         op_linear_quantizer_config: Optional[Dict] = None,
         pass_names: Optional[List[str]] = None,
+        quantize_embedding_tables: bool = False,
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py]: The option is named for embedding tables but it covers every gather whose table is a float constant. That is wider than it sounds. A rotary position cache is two float buffers read by index, and Core ML lowers that to gather, so turning this on quantizes those too. I checked it on a module with no embedding layer at all and both caches were picked up, with about 0.004 absolute error on values between minus one and one.

The comment inside the conversion helper does say this. The problem is that nothing a caller reads says it. The parameter name, the spec key and both new helper docstrings all still say embedding tables, and the partitioner docs page that lists the available compile spec options was not updated.

One more reason to settle it now rather than later. The key string goes into the compile specs that are written into the exported file, so once this ships the spelling is carried by every model produced with it, and renaming then means supporting two spellings.

Either narrow the selection to real embedding lookups, or rename to something like quantize_gather_tables and put the real scope in both docstrings and on the docs page.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 5]
```diff
@@ -67,12 +68,44 @@ def _gathers_sharing_a_weight(mlmodel: ct.models.MLModel) -> List[str]:
...
+            if op.op_type != "gather":
+                continue
+            table = op.inputs.get("x")
+            source = getattr(table, "op", None)
+            if source is None or source.op_type != "const":
+                continue
+            value = getattr(table, "val", None)
+            if value is None:
+                continue
+            if np.issubdtype(np.asarray(value).dtype, np.floating):
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py:110]: The float table filter has no test. All three new tests build a single float embedding, so the dtype check could be deleted and every one of them would still pass. That is the part of the change the last round asked for, so it is the part most worth pinning.

It is cheap to add. Register an integer buffer, say a small uint8 table, index it with the same ids, and lower with the opt in on, asserting the lowering still succeeds. A model with one float embedding and one integer table would cover both halves in one test.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 6]
```diff
@@ -615,11 +684,25 @@ def _convert_to_mlmodel(
             # embedding is one constant feeding both a gather and a linear, and coremltools
             # refuses to compress a constant its consumers disagree about, so opting the
             # gather out there does not skip the table, it fails the whole lowering.
-            tied = _gathers_sharing_a_weight(mlmodel)
+            #
+            # quantize_embedding_tables opts the float tables back in, for models whose
+            # table is most of their weight and which have measured that compressing it is
+            # worth it. It stays off by default: the table is the one weight an embedding
+            # model's output quality rests on most directly, so opting in belongs with the
+            # caller who can measure the result. It applies to every gather with a float
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py:725]: The description still matches the first version of this change. It says two tests and there are three. It describes the option as covering embedding tables, which is no longer what the code does. And the size, latency and cosine numbers were measured before the last commit changed how the opt-in picks gathers, so it would help to say whether they still hold or to re measure them. The test only checks that one file is smaller than the other, so the exact byte counts are not pinned by anything.

The release notes label is also still missing and the bot has already flagged it. This adds an option that export users will call, so it wants an area label rather than none.

**Comment by msluszniak** [comment]
@shoumikhin Could look again, I addressed your comments.

File: backends/apple/coreml/test/test_coreml_partitioner.py [hunk 7]
```diff
@@ -473,6 +473,131 @@ def forward(self, ids):
...
+        buffer still lowers, and still gets smaller, which is what pins the dtype filter:
+        naming the integer table would abort the conversion instead.
+        """
+
+        class MixedTables(torch.nn.Module):
+            def __init__(self, vocab=4096, dim=128):
+                super().__init__()
+                self.embedding = torch.nn.Embedding(vocab, dim)
+                self.register_buffer(
+                    "codebook", torch.randint(0, 255, (vocab,), dtype=torch.uint8)
```
Comment (shoumikhin) [inline backends/apple/coreml/test/test_coreml_partitioner.py]: This test cannot catch the filter being removed. The codebook is a one dimensional table, and the quantizer skips one dimensional tensors on its own, so naming it is a silent no op either way. The test passes with the dtype check deleted. Please use a two dimensional integer table instead, where naming it aborts the conversion and the test proves the filter works.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 8]
```diff
@@ -67,12 +68,47 @@ def _gathers_sharing_a_weight(mlmodel: ct.models.MLModel) -> List[str]:
...
 
+def _gathers_with_a_float_table(mlmodel: ct.models.MLModel) -> List[str]:
+    """
+    Names of gather ops whose table is a floating point constant.
+
+    This is every constant-table index read, not only embedding layers. Core ML lowers
+    `index_select` and plain indexing to `gather` as well, so a float buffer such as a
+    rotary position cache is named here too and gets compressed by the opt-in.
+
+    Integer tables are skipped: a uint8 codebook is not something the linear quantizer
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py]: Small correction to the rationale in the new helper docstring, and in the same sentence on the docs page and in the description. An integer table does not always abort the conversion. I ran the quantizer directly: a two dimensional integer table raises, a one dimensional one is just skipped. The filter is worth keeping, but naming the real trigger would help, because otherwise someone may drop the filter later, see the example still pass, and think it was dead code.

File: docs/source/backends/coreml/coreml-partitioner.md [hunk 9]
```diff
@@ -20,6 +20,8 @@ A list of `CompileSpec`s is constructed with [`CoreMLBackend.generate_compile_sp
...
 - `compute_precision`: The compute precision used by Core ML (`coremltools.precision.FLOAT16` or `coremltools.precision.FLOAT32`).  The default value is `coremltools.precision.FLOAT16`.  Note that the compute precision is applied no matter what dtype is specified in the exported PyTorch model.  For example, an FP32 PyTorch model will be converted to FP16 when delegating to the Core ML backend by default.  Also note that the ANE only supports FP16 precision.
+- `op_linear_quantizer_config`: Post-training weight compression applied by coremltools, as a dict (for example `{"mode": "linear_symmetric", "dtype": "int8", "granularity": "per_channel"}`).  By default this skips `gather` ops, so embedding and other lookup tables are left uncompressed.
```
Comment (shoumikhin) [inline docs/source/backends/coreml/coreml-partitioner.md]: The two new lines are both a little too absolute about the default. The first says gather ops are skipped by default so lookup tables stay uncompressed. That is not true when the table is also some other op's weight: the existing tied weight path already names that gather, and Core ML Tools applies a name before a type, so the table is compressed. A tied embedding and output projection is the normal shape for a small language model, and there is already a test in this file named for that case. The second says integer tables are always skipped. That holds for the new opt in, which checks the dtype, but the tied weight helper has no dtype check, so an integer table shared with another op is still named. Suggest scoping both sentences: the default skips gather tables except where the table is shared with another op, and it is the opt in that skips integer tables.

**Review by shoumikhin — COMMENTED** [review]
Automated review of this change, posted by request. The gap is small and sits in the documentation and the tests rather than in the code path.

Two things matter most. The default exemption and the new option both key on a single Core ML op type, while Core ML has three gather shaped ops, so the documented default does not hold for two of them. And the test named for the integer table filter passes with that filter deleted, because its codebook is one dimensional and the quantizer skips one dimensional tables before it reads the dtype.

The default path is byte identical with the option off. Everything here was reproduced against coremltools 9.0 on Linux x86_64, converting to MIL rather than running a Core ML model, so nothing was executed on Apple hardware. Happy to share the reproduction steps for any of these.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 10]
```diff
@@ -615,11 +695,23 @@ def _convert_to_mlmodel(
...
+            # model's output quality rests on most directly, so opting in belongs with the
+            # caller who can measure the result. It applies to every gather with a float
+            # table, not only to embeddings, because Core ML lowers index_select to gather
+            # too and the two are indistinguishable here.
+            configured = set(_gathers_sharing_a_weight(mlmodel))
+            if CoreMLBackend.quantize_gather_tables_from_compile_specs(compile_specs):
+                configured |= set(_gathers_with_a_float_table(mlmodel))
             config = cto.coreml.OptimizationConfig(
                 global_config=op_linear_quantizer_config,
                 op_type_configs={"gather": None},
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py:733]: The default exemption and the new option both key on one Core ML op type. Core ML has three gather shaped ops. A table read with torch.gather lowers to gather_along_axis, and a table read with two index tensors lowers to gather_nd. Both of those are compressed today whether or not the option is set. So the new docs line, which says lookup tables are left uncompressed by default, is not true for two of the three. Either handle the other two op types in the exemption and in the new helper, or say plainly that this covers single axis gather only. As it stands the option promises control it does not have, and the docs promise a default that does not hold.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 11]
```diff
@@ -397,9 +467,13 @@ def generate_compile_specs(
         model_type: MODEL_TYPE = MODEL_TYPE.MODEL,
         op_linear_quantizer_config: Optional[Dict] = None,
         pass_names: Optional[List[str]] = None,
+        quantize_gather_tables: bool = False,
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py:484]: This exact bug already happened once in this change. Please add a keyword only marker before the new flag, so the next added option cannot silently steal a positional slot again. The partitioner next door already uses one.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 12]
```diff
@@ -277,6 +313,40 @@ def op_linear_quantizer_config_from_compile_specs(
...
+        Returns whether float gather tables opt in to op_linear_quantizer_config.
+        Defaults to False, which is the behaviour of every model lowered before this
+        spec existed.
+
+        Opting in covers every gather reading a float constant, not only embedding
+        layers. See generate_quantize_gather_tables_compile_spec.
+        """
+        for compile_spec in compile_specs:
+            if compile_spec.key == COMPILE_SPEC_KEYS.QUANTIZE_GATHER_TABLES.value:
+                return compile_spec.value.decode("utf-8") == "True"
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py:360]: If a caller passes a truthy value that is not exactly True, like 1 or the string true from a config file, this emits a spec that reads back as False. The flag looks set but the tables stay uncompressed, with no error anywhere. Please normalize with bool when building the spec, so every truthy input spells True.

File: backends/apple/coreml/compiler/coreml_preprocess.py [hunk 13]
```diff
@@ -424,6 +498,12 @@ def generate_compile_specs(
             compile_specs.append(
                 CoreMLBackend.generate_pass_pipeline_compile_spec(pass_names)
             )
+        if quantize_gather_tables:
```
Comment (shoumikhin) [inline backends/apple/coreml/compiler/coreml_preprocess.py:516]: This flag does nothing on its own. The conversion code only reads it when a quantizer config is also set. Please warn when the flag is set without the config, or say so in the docs, so a caller who sets only the flag knows why nothing changed.

**Comment by msluszniak** [comment]
All addressed in 8a51b909. On the three gather-shaped ops I documented the scope rather than widening the exemption, since that would stop compressing tables existing models compress today; happy to do it here instead if you prefer.
